### PR TITLE
Styleguide lists

### LINKED ISSUE #892: Styleguide: Lists
Document Lists used within the app in a styleguide section

Includes List Items and Cards:

screen shot 2016-12-30 at 6 41 47 pm
screen shot 2016-12-30 at 6 41 57 pm

Lists that don't follow the current pattern should be discussed, example:
screen shot 2016-12-30 at 6 43 02 pm

If any types are missing, please comment with example.

Add:

- `<Section name="lists" title="Lists"></Section>`
- define a `Lists.scss`
- define a `Lists.js` in `/styleguide/components/sections`

Styleguide Index: #884
reference #877

### LINKED ISSUE #903: Styleguide: List Actions
** Needs discussion/research **

Styleguide Index: #884
reference #877

### PR DESCRIPTION
**Pull request by mparke**
closes #892 
closes #903 

screen shot 2017-03-16 at 10 26 54 am

### CODE REVIEW

File: src/linodes/layouts/IndexPage.js [hunk 1]
```diff
@@ -108,50 +118,71 @@ export class IndexPage extends Component {
...
-        <tbody>
-          {groups}
-        </tbody>
-      </table>
+      <div>
+        {groups.map(function (group, index) {
+          return (
+            <div key={index}>
+              <div className="GroupLabel">{group.name}</div>
+              <Table
```
Comment (mparke) [inline src/linodes/layouts/IndexPage.js]: Example of the table interface

File: src/linodes/layouts/IndexPage.js [hunk 2]
```diff
@@ -108,50 +118,71 @@ export class IndexPage extends Component {
...
-        ), lg => lg.group
-      ), this.renderGroup);
-
+    const groups = _.sortBy(
+      _.map(_.groupBy(Object.values(linodes), l => l.group), (_linodes, _group) => {
+        return {
+          name: _group,
+          // TODO: don't redefine these for each
+          columns: [
+            { cellComponent: CheckboxCell },
```
Comment (mparke) [inline src/linodes/layouts/IndexPage.js]: example of custom cell rendering

File: src/linodes/layouts/IndexPage.js (hunk 1 above)
Comment (eatonphil) [inline src/linodes/layouts/IndexPage.js]: -1 for breaking out groups into different tables if this does not also make sure that tables in the same group all share the same column widths when columns widths can be different.

**Comment by eatonphil** [comment]
I was hoping to see changes to the configs table too. Is that coming?

**Comment by mparke** [comment]
TODO remaining:

- [ ] Node balancers list
- [ ] Authorized application scopes list
- [ ] Linode configs list
- [ ] Linode backups list
- [ ] Tests
- [ ] Linting

File: src/components/tables/Table.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        <TableHeaderRow columns={columns} />
+      </thead>
+    );
+  } else {
+    tableHeader = (<thead></thead>);
+  }
+
+  let tableContent;
+  if (!data.length) {
+    tableContent = (<p>No records found.</p>);
```
Comment (eatonphil) [inline src/components/tables/Table.js]: Not a big deal, but this is probably not the default message we'd want.

File: src/components/tables/cells/AnchorCell.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  } = column;
+
+  return (
+    <TableCell column={column} record={record}>
+      <a
+        href={hrefFn ? hrefFn(record) : null}
+        className="Table-rowLabel"
+        onClick={onClick}
+        title={record[titleKey]}
+      >
```
Comment (eatonphil) [inline src/components/tables/cells/AnchorCell.js]: I don't think we want to be using anchor tags directly anywhere in the source. Use Link instead?

File: src/components/tables/cells/IPRdnsCell.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import TableCell from './TableCell';
+
+
+export default function IPAddressCell(props) {
+  const { column, record, ipKey } = props;
+
+  const ip = ipKey ? record[ipKey] : record;
+  let str = `${ip.address}`;
+  if (ip.rdns) {
+    str += `(${ip.rdns})`;
```
Comment (eatonphil) [inline src/components/tables/cells/IPRdnsCell.js]: Missing a space here before the (?

File: src/linodes/linode/networking/components/IPList.js [hunk 6]
```diff
@@ -1,32 +1,26 @@
...
-        keys={['address', '']}
-        rows={rows}
+      <Table
+        columns={[
+          { cellComponent: CheckboxCell, onChange: (ip) => { onChange(ip.address); } },
+          { cellComponent: IPRdnsCell },
+          { dataKey: 'address', label: 'IP Address' },
+        ]}
+        data={transferableIps}
+        selected={checked}
```
Comment (eatonphil) [inline src/linodes/linode/networking/components/IPList.js]: This used to be a secondary table?

File: src/linodes/linode/networking/components/IPSharing.js [hunk 7]
```diff
@@ -104,10 +104,20 @@ export default class IPSharing extends Component {
         </p>
         <Form onSubmit={this.onSubmit}>
           <FormGroup>
-            <SecondaryTable
-              labels={['IP Address', 'Linode', '']}
-              keys={['address', 'linode', '']}
-              rows={rows}
+            <Table
```
Comment (eatonphil) [inline src/linodes/linode/networking/components/IPSharing.js]: Used to be a secondary table?

File: src/linodes/linode/networking/layouts/ReverseDNSPage.js [hunk 8]
```diff
@@ -76,10 +79,9 @@ export class ReverseDNSPage extends Component {
 
     return (
       <Card title="Reverse DNS">
-        <SecondaryTable
-          labels={labels}
-          keys={keys}
-          rows={records}
+        <Table
+          columns={columns}
```
Comment (eatonphil) [inline src/linodes/linode/networking/layouts/ReverseDNSPage.js]: Used to be a secondary table?

File: src/nodebalancers/nodebalancer/layouts/IndexPage.js [hunk 9]
```diff
@@ -49,41 +48,33 @@ export class IndexPage extends Component {
...
-        edit: <Button>Edit</Button>,
       };
     });
 
     return (
-      <SecondaryTable
-        labels={labels}
-        keys={keys}
-        rows={newConfigs}
+      <Table
```
Comment (eatonphil) [inline src/nodebalancers/nodebalancer/layouts/IndexPage.js]: Used to be a secondary table?

File: src/components/tables/Table.js (hunk 3 above)
Comment (na3d) [inline src/components/tables/Table.js]: Default is dependent on the usage, no?

File: src/components/tables/cells/AnchorCell.js (hunk 4 above)
Comment (na3d) [inline src/components/tables/cells/AnchorCell.js]: I think I have seen anchors used elsewhere, we may need to go through and do some clean up.

File: src/linodes/linode/networking/components/IPSharing.js (hunk 7 above)
Comment (mparke) [inline src/linodes/linode/networking/components/IPSharing.js]: fixed

File: src/nodebalancers/nodebalancer/layouts/IndexPage.js (hunk 9 above)
Comment (mparke) [inline src/nodebalancers/nodebalancer/layouts/IndexPage.js]: fixed

**Comment by eatonphil** [comment]
Overall looks good. The OAuth scopes are rendered a little funny now though:

screen shot 2017-03-17 at 11 52 00 am

* [ ] .OAuthScopes used to have a padding top 15px
* [ ] The scope title used to be capitalized

**Comment by eatonphil** [comment]
Some padding issues with the buttons here and the table headers should not be bolded:

screen shot 2017-03-17 at 11 54 29 am

**Comment by eatonphil** [comment]
But again, -1 to reverting to multiple tables in a group:

screen shot 2017-03-17 at 11 57 36 am

**Comment by mparke** [comment]
@eatonphil re: groups and tables, they need to be separate at this stage, forcing group definitions into a table is additional complexity. It is 1 table / group.

**Comment by mparke** [comment]
will see if I can fix some of the css problems

**Comment by mparke** [comment]
Fixed some css problems, rebased, fixed tests and linting

screen shot 2017-03-17 at 2 14 25 pm
screen shot 2017-03-17 at 2 14 43 pm
screen shot 2017-03-17 at 2 14 54 pm

**Comment by mparke** [comment]
Follow ups from TODOs:

https://github.com/linode/manager/issues/1410
https://github.com/linode/manager/issues/1411
https://github.com/linode/manager/issues/1412
https://github.com/linode/manager/issues/1413
https://github.com/linode/manager/issues/1414
https://github.com/linode/manager/issues/1415
https://github.com/linode/manager/issues/1416
https://github.com/linode/manager/issues/1417
https://github.com/linode/manager/issues/1418
https://github.com/linode/manager/issues/1419
https://github.com/linode/manager/issues/1420

@na3d

**Comment by mparke** [comment]
and one more https://github.com/linode/manager/issues/1421

**Pull request merged**